src/sparse_blas/backends/mkl_common/mkl_spmv.cxx

```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+void spmv_buffer_size(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                      oneapi::mkl::sparse::matrix_view A_view,
+                      oneapi::mkl::sparse::matrix_handle_t A_handle,
+                      oneapi::mkl::sparse::dense_vector_handle_t x_handle, const void *beta,
+                      oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                      oneapi::mkl::sparse::spmv_alg /*alg*/,
+                      oneapi::mkl::sparse::spmv_descr_t /*spmv_descr*/,
+                      std::size_t &temp_buffer_size) {
+    // TODO: Add support for external workspace once the close-source oneMKL backend supports it.
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     is_beta_host_accessible);
+    temp_buffer_size = 0;
+}
+
+void spmv_optimize(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                   oneapi::mkl::sparse::matrix_view A_view,
+                   oneapi::mkl::sparse::matrix_handle_t A_handle,
+                   oneapi::mkl::sparse::dense_vector_handle_t x_handle, const void *beta,
+                   oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                   oneapi::mkl::sparse::spmv_alg alg,
+                   oneapi::mkl::sparse::spmv_descr_t /*spmv_descr*/,
+                   sycl::buffer<std::uint8_t, 1> /*workspace*/) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     is_beta_host_accessible);
+    auto internal_A_handle = detail::get_internal_handle(A_handle);
+    if (!internal_A_handle->all_use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    if (alg == oneapi::mkl::sparse::spmv_alg::no_optimize_alg) {
+        return;
+    }
+    sycl::event event;
+    internal_A_handle->can_be_reset = false;
+    if (A_view.type_view == matrix_descr::triangular) {
+        event = oneapi::mkl::sparse::optimize_trmv(queue, A_view.uplo_view, opA, A_view.diag_view,
+                                                   internal_A_handle->backend_handle);
+    }
+    else if (A_view.type_view == matrix_descr::symmetric ||
+             A_view.type_view == matrix_descr::hermitian) {
+        // No optimize_symv currently
+        return;
+    }
+    else {
+        event = oneapi::mkl::sparse::optimize_gemv(queue, opA, internal_A_handle->backend_handle);
+    }
+    // spmv_optimize is not asynchronous for buffers as the backend optimize functions don't take buffers.
+    event.wait_and_throw();
```

Comment:
You're right, I think at the time I was worried this would not work as expected because these functions don't take the buffers for the x and y handle. I've removed the event and the wait in https://github.com/oneapi-src/oneMKL/pull/500/commits/d76ca037dca01cf0c53959d3685665e928f08cf9